### PR TITLE
Stop allocating CASE buffer based on wire sizes

### LINKED ISSUE #8924: CASE allocates memory based on wire-provided sizes without sound bounds
#### Problem
In CASESession.cpp, there exists several usage of the following pattern:

`VerifyOrExit(msg_R2_Encrypted.Alloc(tlvReader.GetLength()), err = CHIP_ERROR_NO_MEMORY);`

This assumes in many places that `tlvReader.GetLength()` is not a huge value (or contextually large value), even though only certain values are permissible.

 #### Proposed Solution

All usage of `tlvReader.GetLength()` for allocations should be audited and each one should be validated against a valid upper bound for the size of the value being read, within practical limits.

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
- CASE was using direct lengths of TLV values to allocate heap values,
  without validations.
Fixes #8924

#### Change overview
- Added validations on lengths of Sigma2/3 TBE, with some room to grow
- Updated tag order validation logic to be more explicit and easier to
  read.

#### Testing
- Cert tests still pass
- Unit tests still pass

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

File: src/protocols/secure_channel/CASESession.cpp [hunk 1]
```diff
@@ -906,11 +932,18 @@ CHIP_ERROR CASESession::HandleSigma2(System::PacketBufferHandle && msg)
     SuccessOrExit(err = mCommissioningHash.AddData(ByteSpan{ buf, buflen }));
 
     // Generate decrypted data
-    SuccessOrExit(err = tlvReader.Next());
-    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
-    VerifyOrExit(msg_R2_Encrypted.Alloc(tlvReader.GetLength()), err = CHIP_ERROR_NO_MEMORY);
+    SuccessOrExit(err = tlvReader.Next(TLV::kTLVType_ByteString, TLV::ContextTag(kTag_Sigma2_Encrypted2)));
+
+    max_msg_r2_signed_enc_len =
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASESession.cpp:937]: Could it be constexpr declared here, or do the gotos make that unhappy?

File: src/protocols/secure_channel/CASESession.cpp [hunk 2]
```diff
@@ -1144,20 +1180,25 @@ CHIP_ERROR CASESession::HandleSigma3(System::PacketBufferHandle && msg)
...
 
     tlvReader.Init(std::move(msg));
     SuccessOrExit(err = tlvReader.Next(containerType, TLV::AnonymousTag()));
     SuccessOrExit(err = tlvReader.EnterContainer(containerType));
 
     // Fetch encrypted data
-    SuccessOrExit(err = tlvReader.Next());
-    VerifyOrExit(TLV::TagNumFromTag(tlvReader.GetTag()) == ++decodeTagIdSeq, err = CHIP_ERROR_INVALID_TLV_TAG);
-    VerifyOrExit(msg_R3_Encrypted.Alloc(tlvReader.GetLength()), err = CHIP_ERROR_NO_MEMORY);
+    max_msg_r3_signed_enc_len = TLV::EstimateStructOverhead(Credentials::kMaxCHIPCertLength, Credentials::kMaxCHIPCertLength,
```
Comment (bzbarsky-apple) [inline src/protocols/secure_channel/CASESession.cpp:1190]: Again, could we make this constexpr?

File: src/protocols/secure_channel/CASESession.cpp (hunk 1 above)
Comment (tcarmelveilleux) [reply src/protocols/secure_channel/CASESession.cpp:937]: The gotos make that unhappy :(

**Pull request merged**